### ISSUE TITLE
PJRT CUDA Plugin and Custom Call extensions

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#21807]
### TLDR
When registering custom calls with XLA, if jaxlib is installed with the old 'jaxlib with cuda' mechanism it works correctly, but if jaxlib is installed with the new pjrt cuda backend the custom calls are not found. 

### Description

Hello, I am one of the developers of [mpi4jax](https://github.com/mpi4jax/mpi4jax), a plugin that provides some new primitives to call MPI inside of jax.jitted functions.

To do so, we register several [custom_call_targets](https://github.com/mpi4jax/mpi4jax/blob/93f9a1fd0b092535486dbef981a71e261a340442/mpi4jax/_src/xla_bridge/__init__.py#L35) with xla by calling
```python
    for name, fn in mpi_xla_bridge_cuda.custom_call_targets.items():
        xla_client.register_custom_call_target(name, fn, platform="CUDA")
```
where the [functions are wrapped in a PyCapsule](https://github.com/mpi4jax/mpi4jax/blob/93f9a1fd0b092535486dbef981a71e261a340442/mpi4jax/_src/xla_bridge/mpi_xla_bridge_xpu.pyx#L48).

This worked well for many years, and it still does if one installs jax using the former mechanism which had a separately built jaxlib with cuda support, by doing
```python
pip install -U 'jax[cuda12_pip]' -f https://....
```

however, if I install the new pjrt plugin by doing `pip install -U 'jax[cuda12]'` I get the following error, which suggests I did not register correctly the custom calls.

```python

    def test_allgather_jit():
        from mpi4jax import allgather
    
        arr = jnp.ones((3, 2)) * rank
    
>       res = jax.jit(lambda x: allgather(x)[0])(arr)
E       jaxlib.xla_extension.XlaRuntimeError: UNIMPLEMENTED: No registered implementation for custom call to mpi_allgather for platform CUDA
E       --------------------
E       For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

Are we supposed to register the custom calls in a different way when jaxlib is installed with the pjrt runtime?

### System info (python version, jaxlib version, accelerator, etc.)

```
+ python -c 'import jax; jax.print_environment_info()'
jax:    0.4.29
jaxlib: 0.4.29
numpy:  1.26.4
python: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='r3i5n6', release='5.14.0-284.55.1.el9_2.x86_64', version='#1 SMP PREEMPT_DYNAMIC Mon Feb 19 16:57:59 EST 2024', machine='x86_64')


$ nvidia-smi
Tue Jun 11 17:11:59 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla V100-SXM2-16GB           On  |   00000000:1C:00.0 Off |                    0 |
| N/A   45C    P0             59W /  300W |     311MiB /  16384MiB |      1%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A     97533      C   python                                        308MiB |
+-----------------------------------------------------------------------------------------+

```

**Comment by jyingl3** [comment]
Is there a more specific repro instructions? I tried to install mpi4jax following https://github.com/mpi4jax/mpi4jax?tab=readme-ov-file#installation but failed with `mpi.h: No such file or directory`.

mpi4jax/_src/xla_bridge/__init__.py runs when mpi4jax is imported, is that correct? The failed test is like `import jax; import mpi4jax;` and then `test_allgather_jit`?

**Comment by PhilipVinc** [comment]
@jyingl3 I've updated my answer above. Let me know if you need any help in checking this!
Do know that the CPU version of the code runs fine, so you need to make sure to run this on GPUs.

About your second point...
> mpi4jax/_src/xla_bridge/init.py runs when mpi4jax is imported, is that correct? The failed test is like import jax; import mpi4jax; and then test_allgather_jit?

Yes, `mpi4jax/_src/xla_bridge/init.py ` runs when `mpi4jax` is imported, though I believe that jax processes those registrations only when it initialises the backends.
You need to run an arbitrary mpi4jax function to see the error arise. 
Something as simple as our sample
```python
from mpi4py import MPI
import jax
import jax.numpy as jnp
import mpi4jax

comm = MPI.COMM_WORLD
rank = comm.Get_rank()

@jax.jit
def foo(arr):
   arr = arr + rank
   arr_sum, _ = mpi4jax.allreduce(arr, op=MPI.SUM, comm=comm)
   return arr_sum

a = jnp.zeros((3, 3))
result = foo(a)

if rank == 0:
   print(result)
```
will trigger the error.

**Comment by PhilipVinc** [comment]
> BTW does it work with previous JAX release of 'jax[cuda12]' ?

No. It only works if you use the legacy install method `jax[cuda12_local/pip] -f https://...`

> Now I got error when running pip install --no-build-isolation mpi4jax:

Ops, you broke our `--no-build-isolation` installation. 
Just install `mpi4py` along with Cython!

**Comment by PhilipVinc** [comment]
Updated instructions 
```python
git clone https://github.com/mpi4jax/mpi4jax
cd mpi4jax

pip install --upgrade 'jax[cuda12]' 
pip install cython mpi4py
pip install --no-build-isolation mpi4jax

pytest tests/collective_ops/test_allreduce.py
```

**Comment by jyingl3** [comment]
Thanks! The installation works now, but I got the error below when I run `import mpi4jax`
```
ImportError: cannot import name 'mpi_xla_bridge' from partially initialized module 'mpi4jax._src.xla_bridge' (most likely due to a circular import) (/home/jieying/mpi4jax/mpi4jax/_src/xla_bridge/__init__.py)
```
Any thoughts? Thanks!

**Comment by PhilipVinc** [comment]
Can you share the full stack trace? This should also not happen...

**Comment by jyingl3** [comment]
```
>>> import mpi4jax
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jieying/mpi4jax/mpi4jax/__init__.py", line 9, in <module>
    from ._src import (  # noqa: E402
  File "/home/jieying/mpi4jax/mpi4jax/_src/__init__.py", line 11, in <module>
    from . import xla_bridge  # noqa: E402
  File "/home/jieying/mpi4jax/mpi4jax/_src/xla_bridge/__init__.py", line 4, in <module>
    from . import mpi_xla_bridge
ImportError: cannot import name 'mpi_xla_bridge' from partially initialized module 'mpi4jax._src.xla_bridge' (most likely due to a circular import) (/home/jieying/mpi4jax/mpi4jax/_src/xla_bridge/__init__.py)
```

Is it related to my environment setup (installed jax, etc.)

**Comment by PhilipVinc** [comment]
Can you do a `ls home/jieying/mpi4jax/mpi4jax/_src/xla_bridge` ?

Did you install editable or not? 

I think that it's picking up the locally available mpi4jax folder but you might have installed it non locally, so the `mpi_xla_bridge.*.so` is not in `"/home/jieying/mpi4jax/mpi4jax/_src/xla_bridge/"` but in your env depot.

To fix it, either install editable with
```
pip install --no-build-isolation -e .
```
or delete the mpi4jax folder there so he does not pick it up.

**Comment by jyingl3** [comment]
Thanks so much! I can reproduce now. Will get back when I find out what is wrong.

**Comment by jyingl3** [comment]
The issue is because register_custom_call_target in cuda plugin does not support bytes as the function name. Sent out https://github.com/google/jax/pull/21814/ to fix it.

**Comment by PhilipVinc** [comment]
Thanks a lot for looking into this! 

Is there a hot fix on our side possible in the meantime? 
I tried to change [this line](https://github.com/mpi4jax/mpi4jax/blob/93f9a1fd0b092535486dbef981a71e261a340442/mpi4jax/_src/xla_bridge/__init__.py#L37) to `xla_client.register_custom_call_target(str(name), fn, platform="CUDA")` but (on the current jax release) it gives out the same error as before, which surprises me...

**Comment by PhilipVinc** [comment]
Ah, no, I got it.
I should do `name.decode('utf-8')`!

Thanks a lot for the help!

**Comment by jyingl3** [comment]
Sorry for the hiccup and thank you for doing the hot fix on your side!
